Systematics deprojection

ellison-j opened the pull request:
Added systematics deprojection for galaxy number counts in TXTwoPointFourier.
See https://github.com/LSSTDESC/TXPipe/issues/98.

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -64,6 +64,9 @@ class TXTwoPointFourier(PipelineStage):
         "flip_g1": False,
         "flip_g2": False,
         "cache_dir": '',
+        "syst_nc": False,
```
Could you give these a slightly more descriptive name, e.g. `deproject_clustering_sys`?  And also remove the lensing option for now as it doesn't do anything yet.

ellison-j replied:
ok

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -221,11 +223,85 @@ def load_maps(self):
+            s_maps = []
+            systmaps_dir = self.config['systmaps_nc_dir']
+            systmaps_path = pathlib.Path(systmaps_dir)
+            try:
```
should this try statement be inside the loop so that if one map doesn't work the others can still deproject?

ellison-j replied:
This was here to catch the problem of no map files in the given directory. If one of the maps is a .fits file that is not a proper healpix maps file `healpy.read_map` will give an error and the job will stop. So I am not sure about moving it.

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -221,11 +223,85 @@ def load_maps(self):
+            try:
+                for systmap in systmaps_path.iterdir():
+                    if systmap.is_file():
+                        systmap_file = str(systmap)
```
maybe also check if the filename ends with .fits or .hs here and only use it if so?

ellison-j replied:
I'll check that it is a .fits file since at present only HEALPix is implemented (we can add healsparse later when TXPipe supports it).

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -221,11 +223,85 @@ def load_maps(self):
+
+                        # normalize map for Namaster
+                        # set pixel values to value/mean - 1
+                        syst_map_mask = syst_map != healpy.UNSEEN
```
are there typically any nans in these maps to worry about?

ellison-j replied:
There should not be any NaN's in the HELAPix map as far as I know.

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -221,11 +223,85 @@ def load_maps(self):
+                            syst_map = syst_map / mean - 1
+                            print('Syst map', systmap_file, 'normalized value at ra,dec = 55,-30: ', syst_map[ipix])
+                        
+                        s_maps.append(syst_map)
```
could you also set something like `self.config[f'clustering_deproject_{n_systmaps}'] = systmap_file` so it gets recorded in the output file provenance?

ellison-j replied:
OK, but I don't understand how the provenance works. Do I just add that line and let the provenance code do the rest or do I have to do something more?

joezuntz replied:
Yep, that's right - everything in the `self.config` dictionary gets put in the output file metadata when it gets opened, automatically.

joezuntz commented on txpipe/twopoint_fourier.py:
```diff
@@ -221,11 +223,85 @@ def load_maps(self):
+            # needed for NaMaster:
+            s_maps_nc = np.array(s_maps).reshape([n_systmaps, 1, npix])
+
+# wl not implemented yet
```
A good question!  Maybe look for pairs of maps?
We could also read a text file of paths to deproject, and look for pairs for shear.  That can wait, though.

ellison-j replied:
Yes, we could use ellipticity e1, e2. But will leave lensing systematics out for now.

ellison-j commented:
Hi @joezuntz I have implemented your suggestions and updated the branch. I am now rejecting any map files that are not valid and using all the others, rather than halting the job if one file is bad.  This is fine I think, but the user needs to check the log file to make sure all maps provided were used.

I also changed the yml file to use only 16 threads since with 64 threads I was running out of memory with 4 maps.

Let me know if you have any more comments.
